WireDomain/Sources/WireDomain/Repositories/Message/MessageLocalStore.swift

```diff
@@ -71,10 +210,337 @@ public final class MessageLocalStore: MessageLocalStoreProtocol {
+    }
+
+    public func deleteMessageForSelf(
+        _ hiddenMessage: MessageHide,
+        in conversation: ZMConversation
+    ) async {
+        await context.perform { [context] in
+            guard conversation.isSelfConversation else {
+                return
+            }
+
+            ZMMessage.remove(
+                remotelyHiddenMessage: hiddenMessage,
+                inContext: context
+            )
+        }
+    }
+
+    public func deleteMessageForEveryone(
+        _ deletedMessage: MessageDelete,
+        in conversation: ZMConversation,
+        senderID: UUID
+    ) async {
+        await context.perform { [context] in
+            ZMMessage.remove(
+                remotelyDeletedMessage: deletedMessage,
+                inConversation: conversation,
+                senderID: senderID,
+                inContext: context
+            )
+        }
+    }
+
+    public func addMessageReaction(
+        _ messageReaction: WireProtos.Reaction,
+        in conversation: ZMConversation,
+        senderID: UUID,
+        date: Date
+    ) async {
+        await context.perform { [context] in
+            ZMMessage.add(
+                reaction: messageReaction,
+                senderID: senderID,
+                conversation: conversation,
+                creationDate: date,
+                inContext: context
+            )
+        }
+    }
+
+    public func updateButtonStates(
+        _ buttonActionConfirmation: ButtonActionConfirmation,
+        in conversation: ZMConversation
+    ) async {
+        await context.perform { [context] in
+            ZMClientMessage.updateButtonStates(
+                withConfirmation: buttonActionConfirmation,
+                forConversation: conversation,
+                inContext: context
+            )
+        }
+    }
+
+    public func editMessage(
+        _ messageEdit: MessageEdit,
+        in conversation: ZMConversation,
+        senderID: UUID,
+        genericMessage: GenericMessage,
+        date: Date
+    ) async {
+        guard let editedMessageID = UUID(
+            uuidString: messageEdit.replacingMessageID
+        ) else {
+            return
+        }
+
+        await context.perform { [self] in
+            guard let editedClientMessage = ZMClientMessage.fetch(
+                withNonce: editedMessageID,
+                for: conversation,
+                in: context
+            ) else {
+                return
+            }
+
+            guard editMessage(
+                messageEdit,
+                clientMessage: editedClientMessage,
+                genericMessage: genericMessage,
+                senderID: senderID,
+                date: date
+            ) else {
+                return
+            }
+
+            editedClientMessage.updateCategoryCache()
+            editedClientMessage.markAsSent()
+        }
+    }
+
     // MARK: - Private
 
+    private func shouldAddMessage(to conversation: ZMConversation, date: Date) -> Bool {
```

Comment:
This seems to have some overlap with `canAddMessage`. Can they be combined?

Reply:
removed in https://github.com/wireapp/wire-ios/commit/244f4d59aa504d7edb51a223ddcf618e4d691b75